feat: added basic preprocess table configuration

arteevraina opened the pull request:
* This PR adds the preprocess table configuration.
* Add a compiler flag when [preprocess.cpp] is found in `manifest` file. (WIP)

milancurcic commented:
Thanks, @arteevraina. Include in this PR also a minimal example package in [https://github.com/fortran-lang/fpm/tree/main/example_packages], e.g. `preprocess_cpp` or similar, that will be used for testing.

arteevraina commented:
@milancurcic Sure.

milancurcic commented on example_packages/preprocess_cpp/src/preprocess_cpp.f90:
```diff
@@ -0,0 +1,10 @@
+
+  public :: say_hello
+contains
+  subroutine say_hello
+    print *, "Hello, preprocess_cpp!"
+  end subroutine say_hello
```
The current example won't test whether the preprocessing worked. Can you add the logic here that would do that?

For example, you could add an `#ifndef` block that encloses some invalid code that would break the compilation if that block wasn't correctly preprocessed:

```fortran
  subroutine say_hello
    print *, "Hello, preprocess_cpp!"
#ifndef TESTMACRO
    break me
#endif
  end subroutine say_hello
```

and `-DTESTMACRO` would need to be added to the flags for the example package build to succeed.

@awvwgk @LKedward does this approach seem appropriate to you?

milancurcic replied:
Sorry, @arteevraina for the delay getting back to you. I think we should leave the macros in the manifest for a separate PR, to keep them small. What I meant by "passing `-DTESTMACRO` to the flags" is simply that a user who builds the example package would need to pass it explicitly to fpm, e.g. `fpm build --flag "-DTESTMACRO" for the example package to successfully build.

Let's keep this PR small but complete. You'll still need to add a few tests (`test/fpm_test/test_manifest.f90` seems the appropriate file to add them to), and update the [manifest reference](https://github.com/fortran-lang/fpm-docs/blob/main/pages/spec/manifest.md) in a separate PR to fpm-docs.

arteevraina replied:
@milancurcic Sure thing.

milancurcic replied:
No, the command should not include `-cpp` because this example package would have `[preprocess.cpp]` in the manifest, so fpm would add `-cpp` to the default flags under the hood.

Yes, src/fpm_compiler.f90 is the place to add definitions to enable the C-preprocessor flag for each of the supported compilers.

arteevraina commented:
@milancurcic I have added the ability to add `cpp` flag and also added the test cases in this PR.
Will open a new PR for docs when this gets merged. Let me know if any changes are required in this PR.

milancurcic commented on src/fpm/manifest/preprocess.f90:
```diff
@@ -0,0 +1,192 @@
+
+      call table%get_keys(list)
+
+      ! An empty table is not allowed
+      if (size(list) == 0) then
+         call syntax_error(error, "No preprocessors defined")
+      end if
```
UI question for @arteevraina and @fortran-lang/fpm: Should just

```toml
[preprocess]
```

be allowed and imply

```toml
[preprocess]
[preprocess.cpp]
```
?

I'm split. On one side, explicit is more clear. On the other, requiring `[preprocess.cpp]` adds the needed boilerplate if one just wants CPP preprocessing everywhere as a sane default, it could be nice to just type `[preprocess]` in the manifest and be done with it, and a sane default philosophy is I think aligned with the original vision for fpm. But it would effectively make CPP the first-class preprocessor, for better or for worse.

arteevraina replied:
I would say being explicit about which preprocessors we want to use while building fortran projects using fpm would make the `manifest` file clear and what goes under the hood. But, if we are going to make `cpp` as first choice preprocessors for fpm-based projects then I think simply `[preprocess]` should be fine. I think more or less this decision depends on how we expect to see fpm in the future. Either it should have a favorite `cpp` preprocessor with allowing other preprocessors as well or just being preprocessor agnostic and leaving it to the user to be explicit.

jvdp1 replied:
I would prefer to be explicit too. For example, if a compiler does not support `cpp`, the default option would fail. Furthermore we don't know if `cpp` will still be the "default"/preferred preprocessor in the future.

milancurcic commented on src/fpm_compiler.f90:
```diff
@@ -376,6 +381,32 @@ subroutine get_debug_compile_flags(id, flags)
+       return
+    end if
+    
+    !> Check if there is preprocess table
```
```suggestion
    !> Check if there is a preprocess table
```

milancurcic reviewed (approved):
Thanks, @arteevraina!

jvdp1 commented on src/fpm/manifest/preprocess.f90:
```diff
@@ -0,0 +1,192 @@
+      integer :: pr, ilink
+      character(len=*), parameter :: fmt = '("#", 1x, a, t30, a)'
+
+      if (present(verbosity)) then
+         pr = verbosity
+      else
+         pr = 1
+      end if
```
`pr` is defined but not used.  Is it aimed to be used later?

arteevraina replied:
Yeah, I forgot to add the basic check using pr. I have added a new commit for this [here](https://github.com/fortran-lang/fpm/pull/715/commits/c5030f1e50fc89ff8be4d01764162854b6203ed6).

jvdp1 commented on src/fpm_compiler.f90:
```diff
@@ -177,6 +179,9 @@ module fpm_compiler
     flag_lfortran_opt = " --fast"
 
+character(*), parameter :: &
+    flag_cpp_preprocessor = " -cpp"
```
Do all compilers support the flag `-cpp` for the integrated preprocessor?
I know that Intel Fortran compiler uses `-fpp` (but may support `-cpp` too). Therefore it might be good to add an option to modify this flag.

arteevraina replied:
Thanks for pointing it out. I think while writing code, I assumed all compilers support `-cpp` for preprocessing.
Do I need to consider these compilers mentioned [here](https://github.com/fortran-lang/fpm/blob/main/src/fpm_compiler.f90#L5) ?

jvdp1 reviewed (commented):
Nice work.
I have a general question. It seems that it is assumed that the CPP preprocessor is always provided by the compiler. Is it the case for all available compilers?

arteevraina commented:
@jvdp1 I have added the logic to modify `cpp preprocessor flag` [here](https://github.com/fortran-lang/fpm/pull/715/commits/1c2f23501a47975d9e9e7262e1267f0684d3d1c8).

zoziha commented on src/fpm_compiler.f90:
```diff
@@ -398,6 +399,18 @@ subroutine set_preprocessor_flags (flags)
+        flag_cpp_preprocessor = ""
+    case(id_caf, id_gcc, id_f95, id_nvhpc)
+        flag_cpp_preprocessor = "-cpp"
+    case(id_intel_classic_nix, id_intel_classic_mac, id_intel_classic_windows, id_intel_llvm_nix, id_intel_llvm_windows, id_nag)
+        flag_cpp_preprocessor = "-fpp"
```
Thanks for your great work @arteevraina. `id_intel_classic_windows` and `id_intel_llvm_windows` should be `/fpp`, and the preprocess option of Intel's Fortran compiler on Windows OS is `/fpp`, not `-fpp`.

PS: It is worth noting that on Linux and Mac systems, most compilers enable preprocessing by default for `.F90` files, but on Windows systems, Intel's compilers do not enable preprocessing **by default** for `.F90` files.

arteevraina replied:
Thanks for reviewing @zoziha. I think if we make changes as per the your comments. I think the final logic should look like this one :

```
    !> Modify the flag_cpp_preprocessor on the basis of the compiler.
    select case(id)
    case default
        flag_cpp_preprocessor = ""
    case(id_caf, id_gcc, id_f95, id_nvhpc)
        flag_cpp_preprocessor = "-cpp"
    case(id_intel_classic_windows, id_intel_llvm_windows)
        flag_cpp_preprocessor = "/fpp"
    case(id_intel_classic_nix, id_intel_classic_mac, id_intel_llvm_nix, id_nag)
        flag_cpp_preprocessor = "-fpp"
    case(id_lfortran)
        flag_cpp_preprocessor = "--cpp"
    end select
```
Can you confirm this ?

jvdp1 reviewed (approved):
Pending remaining comments, it looks good to me!

milancurcic commented:
Yes, I think so, let's move forward. Thanks @arteevraina for the PR and everybody else for the reviews.
